go/heif/heif.go

```diff
@@ -637,6 +754,217 @@ func (img *Image) ScaleImage(width int, height int) (*Image, error) {
 	return &scaled_image, nil
 }
 
+// --- High-level encoding API.
+
+type EncodingOptions struct {
+	options *C.struct_heif_encoding_options
+}
+
+func NewEncodingOptions() (*EncodingOptions, error) {
+	options := &EncodingOptions{
+		options: C.heif_encoding_options_alloc(),
+	}
+	if options.options == nil {
+		return nil, fmt.Errorf("Could not allocate encoding options")
+	}
+
+	runtime.SetFinalizer(options, freeHeifEncodingOptions)
+	return options, nil
+}
+
+func freeHeifEncodingOptions(options *EncodingOptions) {
+	C.heif_encoding_options_free(options.options)
+	options.options = nil
+	runtime.SetFinalizer(options, nil)
+}
+
+func imageFromNRGBA(i *image.NRGBA) (*Image, error) {
+	max := i.Bounds().Max
+	w, h := max.X, max.Y
+
+	hasAlpha := func(i *image.NRGBA) bool {
+		rect := i.Rect
+		if rect.Empty() {
+			return true
+		}
+
+		for y := rect.Min.Y; y < rect.Max.Y; y++ {
+			for x := rect.Min.X; x < rect.Max.X; x++ {
+				o := i.PixOffset(x, y)
+				if i.Pix[o+3] == 0xff {
+					return true
+				}
+			}
+		}
+		return false
+	}
+
+	var cm Chroma
+	var depth int
+	var stride int
+	if hasAlpha(i) {
+		cm = ChromaInterleavedRGBA
+		depth = 32
+		stride = w * 4
+	} else {
+		cm = ChromaInterleavedRGB
+		depth = 64
+		stride = w * 3
+	}
+
+	out, err := NewImage(w, h, ColorspaceRGB, cm)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create image: %v", err)
+	}
+
+	p, err := out.NewPlane(ChannelInterleaved, w, h, depth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add plane: %v", err)
+	}
+	p.SetData([]byte(i.Pix), stride)
+
+	return out, nil
+}
+
+func imageFromGray(i *image.Gray) (*Image, error) {
+	max := i.Bounds().Max
+	w, h := max.X, max.Y
+
+	out, err := NewImage(w, h, ColorspaceYCbCr, ChromaMonochrome)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create image: %v", err)
+	}
+
+	const depth = 8
+	pY, err := out.NewPlane(ChannelY, w, h, depth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add Y plane: %v", err)
+	}
+	pY.SetData([]byte(i.Pix), i.Stride)
+
+	return out, nil
+}
+
+func imageFromYCbCr(i *image.YCbCr) (*Image, error) {
+	max := i.Bounds().Max
+	w, h := max.X, max.Y
+
+	var cm Chroma
+	switch i.SubsampleRatio {
+	case image.YCbCrSubsampleRatio444:
+		cm = Chroma444
+	case image.YCbCrSubsampleRatio422:
+		cm = Chroma422
+	case image.YCbCrSubsampleRatio420:
+		fallthrough
+	default:
+		cm = Chroma420
+	}
+
+	out, err := NewImage(w, h, ColorspaceYCbCr, cm)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create image: %v", err)
+	}
+
+	const depth = 8
+	pY, err := out.NewPlane(ChannelY, w, h, depth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add Y plane: %v", err)
+	}
+	pY.SetData([]byte(i.Y), i.YStride)
+
+	// TODO: Might need to be updated for other SubsampleRatio values.
```

Comment:
This needs to be implemented if you support 4:2:2 and 4:4:4 above.

Reply:
As discussed, I will remove support for subsample ratios other than 4:2:0.